### PR TITLE
Enable people for audio files and add music probing test

### PR DESCRIPTION
**Pull request by MrTimscampi**
**Companion PR on jf-web:** jellyfin/jellyfin-web#2801

**Changes**

Enables the SupportsPeople flag for Audio entities. Currently, we parse and normalize composers, lyricists and other info from the FFProbe results, but they get thrown out by the FFProbeAudioInfo provider due to that flag being set to false.

image

It's not shown in jf-web or other clients currently, but this will allow clients to show more infos on the tracks, and providers to be improved.

This also adds a test for music parsing, so we can ensure all the audio-specific stuff the FFProbe normalizer does works.

Also added are a bunch of new `PersonType`, meant to cover some of the tags we miss currently in audio files: Arranger, Engineer, Mixer and Remixer.

Finally, it adds performer parsing to the ffprobe normalization performed on audio files.

image

It works identically to genres and composers, except it retrieves the role using Regex (The "standard" way performers are set is using `Full Name (role)`. So we match both the name and the role, then capitalize the role based on the current culture, for display purposes).

**Issues**

### CODE REVIEW

File: tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs [hunk 1]
```diff
@@ -71,5 +72,32 @@ public void GetMediaInfo_MusicVideo_Success()
...
+            MediaInfo res = _probeResultNormalizer.GetMediaInfo(internalMediaInfoResult, null, true, "Test Data/Probing/music.flac", MediaProtocol.File);
+
+            Assert.Equal("UP NO MORE", res.Name);
+            Assert.Single(res.Artists);
+            Assert.Equal("TWICE", res.Artists[0]);
+            Assert.Equal("Eyes wide open", res.Album);
+            Assert.Equal(2020, res.ProductionYear);
+            Assert.True(res.PremiereDate.HasValue);
+            Assert.Equal(DateTime.Parse("2020-10-26T00:00Z", DateTimeFormatInfo.CurrentInfo).ToUniversalTime(), res.PremiereDate);
+            Assert.NotEmpty(res.People);
```
Comment (Bond-009) [inline tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs:90]: Check the exact length

File: tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs [hunk 2]
```diff
@@ -71,5 +72,32 @@ public void GetMediaInfo_MusicVideo_Success()
...
+            Assert.Equal("Krysta Youngs", res.People[0].Name);
+            Assert.Equal(PersonType.Composer, res.People[0].Type);
+            Assert.Equal("Julia Ross", res.People[1].Name);
+            Assert.Equal(PersonType.Composer, res.People[1].Type);
+            Assert.Equal("Yiwoomin", res.People[2].Name);
+            Assert.Equal(PersonType.Composer, res.People[2].Type);
+            Assert.Equal("Ji-hyo Park", res.People[3].Name);
+            Assert.Equal(PersonType.Lyricist, res.People[3].Type);
+            Assert.NotEmpty(res.Genres);
```
Comment (Bond-009) [inline tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs:99]: Don't need this bc of the next line

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs (hunk 2 above)
Comment (MrTimscampi) [reply tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs:99]: How so? I thought that `Assert.Same` was testing for identical instances, while `Assert.Equal` checked the contents?

Should I assert each value like above? Is there a neater way to do that? (Both for the genres array and the persons one)

Comment (MrTimscampi) [reply tests/Jellyfin.MediaEncoding.Tests/Probing/ProbeResultNormalizerTests.cs:99]: Right, I see that `Assert.Equal` uses the Equal method of the type, which arrays don't override, so it checks for instance too.

File: MediaBrowser.Model/Entities/PersonType.cs [hunk 3]
```diff
@@ -1,48 +1,68 @@
...
         /// </summary>
         public const string Conductor = "Conductor";
 
         /// <summary>
         /// The lyricist.
         /// </summary>
         public const string Lyricist = "Lyricist";
+
+        /// <summary>
+        /// A person who writes the words to a song or musical.
```
Comment (neopc10) [inline MediaBrowser.Model/Entities/PersonType.cs:49]: "...Reimagines an existing composition, adapting the instruments, voices, rhythms, and tempo to create a new sound for a piece of music."

Comment (MrTimscampi) [reply MediaBrowser.Model/Entities/PersonType.cs:49]: Ah, seems I mis-pasted and pasted the Lyricist one over Arranger. Thanks for helping me notice :)

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 4]
```diff
@@ -1110,7 +1111,27 @@ private void SetAudioInfoFromTags(MediaInfo audio, IReadOnlyDictionary<string, s
                 }
             }
 
-            // Check for writer some music is tagged that way as alternative to composer/lyricist
+            if (tags.TryGetValue("performer", out var performer) && !string.IsNullOrWhiteSpace(performer))
+            {
+                foreach (var person in Split(performer, false))
+                {
+                    Regex pattern = new Regex(@"(?<name>.*) \((?<instrument>.*)\)");
```
Comment (crobibero) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:1118]: Creating a `Regex` is expensive, I would initialize it at the class level

**Review by barronpm — APPROVED** [review]

**Review by crobibero — APPROVED** [review]

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 5]
```diff
@@ -27,7 +28,9 @@ public class ProbeResultNormalizer
 
         private readonly char[] _nameDelimiters = { '/', '|', ';', '\\' };
 
-        private readonly CultureInfo _usCulture = new CultureInfo("en-US");
+        private readonly Regex _performerPattern = new (@"(?<name>.*) \((?<instrument>.*)\)");
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:31]: Now you will create a new `Regex` instance for each ffprobe process, you can make this static instead (it's thread safe)

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs [hunk 6]
```diff
@@ -1110,7 +1113,26 @@ private void SetAudioInfoFromTags(MediaInfo audio, IReadOnlyDictionary<string, s
...
+                    Match match = _performerPattern.Match(person);
+
+                    // If the performer doesn't have any instrument/role associated, it won't match. In that case, chances are it's simply a band name, so we skip it.
+                    if (match.Success)
+                    {
+                        people.Add(new BaseItemPerson
+                        {
+                            Name = match.Groups["name"].Value,
+                            Type = PersonType.Actor,
+                            Role = CultureInfo.CurrentCulture.TextInfo.ToTitleCase(match.Groups["instrument"].Value)
```
Comment (Bond-009) [inline MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:1129]: Personally, I'd leave it like it is

**Review by Bond-009 — CHANGES_REQUESTED** [review]

File: MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs (hunk 6 above)
Comment (MrTimscampi) [reply MediaBrowser.MediaEncoding/Probing/ProbeResultNormalizer.cs:1129]: It's lowercase in what Picard puts in, at least. So it'd look a bit out of place in clients.

**Review by Bond-009 — APPROVED** [review]

**Pull request merged**